src/topology/config/watcher.rs

```diff
@@ -0,0 +1,136 @@
+use crate::Error;
+use notify::{raw_watcher, Op, RawEvent, RecommendedWatcher, RecursiveMode, Watcher};
+use std::{
+    path::{Path, PathBuf},
+    sync::mpsc::{channel, Receiver},
+    thread,
+    time::Duration,
+};
+
+/// Per notify own documentation, it's advised to have delay of more than 30 sec,
+/// so to avoid receiving repetitions of previous events on macOS.
+///
+/// But, config and topology reload logic can handle:
+///  - Invalid config, caused either by user or by data race.
+///  - Frequent changes, caused by user/editor modifying/saving file in small chunks.
+/// so we can use smaller, more responsive delay.
+pub const CONFIG_WATCH_DELAY: std::time::Duration = std::time::Duration::from_secs(1);
+
+const RETRY_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
+
+/// Triggers SIGHUP when file on config_path changes.
+/// Accumulates file changes until no change for given duration has occured.
+/// Has best effort guarante of detecting all file changes from the end of
+/// this function until the main thread stops.
+#[cfg(unix)]
+pub fn config_watcher(config_path: PathBuf, delay: Duration) -> Result<(), Error> {
+    // Create watcher now so not to miss any changes happening between
+    // returning from this function and the thread starting.
+    let mut watcher = create_watcher(&config_path);
+
+    info!("Watching configuration file.");
+
+    thread::spawn(move || loop {
+        if let Some((_, receiver)) = watcher.take() {
+            while let Ok(RawEvent { op: Ok(event), .. }) = receiver.recv() {
+                if event.intersects(Op::CREATE | Op::REMOVE | Op::WRITE | Op::CLOSE_WRITE) {
+                    info!("Configuration file change detected.");
+
+                    // Consume events until delay amount of time has passed since the latest event.
+                    while let Ok(..) = receiver.recv_timeout(delay) {}
+
+                    info!("Configuration file changed.");
+                    raise_sighup();
+                } else {
+                    debug!(message = "Ignoring event", ?event)
+                }
+            }
+        }
+
+        thread::sleep(RETRY_TIMEOUT);
+
+        watcher = create_watcher(&config_path);
+
+        if watcher.is_some() {
+            // Config file could have changed while we weren't watching,
+            // so for a good measure raise SIGHUP and let reload logic
+            // determine if anything changed.
+            info!("Speculating that configuration file has changed.");
+            raise_sighup();
+        }
+    });
+
+    Ok(())
+}
+
+#[cfg(windows)]
+/// Errors on Windows.
+pub fn config_watcher(config_path: PathBuf, delay: Duration) -> Result<(), Error> {
+    Err("Reloading config on Windows isn't currently supported. Related issue https://github.com/timberio/vector/issues/938 .")
```

Comment:
This does not compile locally

![](https://i.imgur.com/03xPkLA.png)

Reply:
Forgot a `.into()` there.